### ISSUE TITLE
IndicBERT-v3-270M: attention_mask not honoured — real-token hidden states change drastically with batch padding

### ISSUE THREAD

**Issue by TanveerKahlon** [AI4Bharat/IndicBERT#20]
### Affected model
`ai4bharat/IndicBERT-v3-270M` (loaded with `trust_remote_code=True`; remote class `BidirectionalGemma3ForCausalLM`).

### Environment
- `transformers` 5.5.0, `torch` 2.10.0
- Reproduced on **CPU, fp32** (so this is not a bf16 / CUDA-kernel artifact)

### Summary
For a bidirectional encoder, a real token's final hidden state should be **invariant to right-padding** when a correct `attention_mask` is supplied. With this model it is not: encoding a short sequence **alone** vs **in a right-padded batch** changes its real-token `last_hidden_state` by **~10× the state RMS**. This indicates real tokens are attending to PAD positions — `attention_mask` is not being applied in the (bidirectional) attention path.

### Reproduction
```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

MODEL_ID = "ai4bharat/IndicBERT-v3-270M"
tok = AutoTokenizer.from_pretrained(MODEL_ID)
model = AutoModelForCausalLM.from_pretrained(
    MODEL_ID, dtype=torch.float32, trust_remote_code=True).eval()

short = "ਅੱਜ ਮੌਸਮ ਵਧੀਆ ਹੈ"                       # any text reproduces it
long  = "ਇਹ ਇੱਕ ਲੰਮਾ ਵਾਕ ਹੈ ਜੋ ਬੈਚ ਵਿੱਚ ਪੈਡਿੰਗ ਬਣਾਉਣ ਲਈ ਵਰਤਿਆ ਜਾਂਦਾ ਹੈ"

with torch.no_grad():
    a  = tok(short, return_tensors="pt")
    ha = model(**a, output_hidden_states=True).hidden_states[-1][0]   # short alone
    b  = tok([short, long], padding=True, return_tensors="pt")        # short right-padded
    hb = model(**b, output_hidden_states=True).hidden_states[-1][0]
    L  = int(a["attention_mask"][0].sum())

print((ha[:L] - hb[:L]).abs().max().item())   # should be ~0
```

### Observed
```
real-token positions: 12
max|alone - padded_in_batch| on REAL tokens = 94.79
state RMS ~= 9.21  ->  relative drift ~= 10.3
```
Expected: `max|diff| ≈ 0` (within fp32 noise). Actual: ~95 (≈1030% of RMS). The magnitude is text-independent; any pair where the shorter sequence gets right-padded reproduces it.

### Why it matters
Any workflow that encodes variable-length inputs in a padded batch gets **batch-composition-dependent** representations:
- representation / embedding caching (identical text → different vectors per batch),
- mean / attention pooling for retrieval or compression,
- distillation / feature extraction.

The output silently depends on which other sequences share the batch.

### Likely cause
The custom bidirectional attention appears to ignore `attention_mask` (or builds a key-padding / position mask that doesn't exclude PAD keys). A causal model hides right-pad tokens behind the causal mask, masking the bug; a **bidirectional** model exposes it.

### Workaround for users today
Encode each sequence **without cross-padding** — batch only equal-length sequences (no PAD tokens), or encode one-at-a-time. In fp32 this makes real-token states batch-invariant to ~2e-4.

**Comment by TanveerKahlon** [comment]
Cross-linked on the HF model page so v3-270M users find it: https://huggingface.co/ai4bharat/IndicBERT-v3-270M/discussions/1
